File: 11/11.go
package main

import (
	"fmt"
	"math/big"
	"os"
	"sort"
	"strconv"
	"strings"
)

func splitOnNewline(lines []string) [][]string {
	blocks := make([][]string, 1)
	for _, line := range lines {
		if line == "" {
			blocks = append(blocks, nil)
		} else {
			blocks[len(blocks)-1] = append(blocks[len(blocks)-1], line)
		}
	}
	if len(blocks[len(blocks)-1]) == 0 {
		return blocks[:len(blocks)-1]
	}
	return blocks
}

type monkey struct {
	items          []*big.Int
	op             func(*big.Int) *big.Int
	divisibleTest  *big.Int
	throwTrue      int
	throwFalse     int
	inspectedCount uint64
}

func check2[T any](v T, err error) T {
	if err != nil {
		panic(err)
	}
	return v
}

func lastInt(s string) int {
	f := strings.Fields(s)
	return check2(strconv.Atoi(f[len(f)-1]))
}

func lastBig(s string) *big.Int {
	f := strings.Fields(s)
	return big.NewInt(check2(strconv.ParseInt(f[len(f)-1], 10, 64)))
}

func parseItems(s string) []*big.Int {
	itemsRaw := strings.Split(strings.TrimSpace(strings.Split(s, ":")[1]), ", ")
	items := make([]*big.Int, len(itemsRaw))
	for i, itemRaw := range itemsRaw {
		items[i] = big.NewInt(check2(strconv.ParseInt(itemRaw, 10, 64)))
	}
	return items
}

func parseOp(s string) func(*big.Int) *big.Int {
	opRaw := strings.Fields(s)
	switch opRaw[len(opRaw)-2] {
	case "+":
		y := lastBig(opRaw[len(opRaw)-1])
		return func(old *big.Int) *big.Int {
			var z big.Int
			return z.Add(old, y)
		}
	case "*":
		if opRaw[len(opRaw)-1] == "old" {
			return func(old *big.Int) *big.Int {
				var z big.Int
				return z.Mul(old, old)
			}
		}
		y := lastBig(opRaw[len(opRaw)-1])
		return func(old *big.Int) *big.Int {
			var z big.Int
			return z.Mul(old, y)
		}
	default:
		panic("internal")
	}
}

func parse(monkeysRaw [][]string) []*monkey {
	monkeys := make([]*monkey, 0)
	for _, monkeyRaw := range monkeysRaw {
		monkeys = append(monkeys, &monkey{
			items:         parseItems(monkeyRaw[1]),
			op:            parseOp(monkeyRaw[2]),
			divisibleTest: lastBig(monkeyRaw[3]),
			throwTrue:     lastInt(monkeyRaw[4]),
			throwFalse:    lastInt(monkeyRaw[5]),
		})
	}
	return monkeys
}

var three = big.NewInt(3)

func round(monkeys []*monkey, decreaseWorryLevel bool) {
	for _, monkey := range monkeys {
		for _, item := range monkey.items {
			newWorryLevel := monkey.op(item)
			if decreaseWorryLevel {
				newWorryLevel.Div(newWorryLevel, three)
			}
			var z big.Int
			if z.Mod(newWorryLevel, monkey.divisibleTest).Sign() == 0 {
				nextItems := &monkeys[monkey.throwTrue].items
				*nextItems = append(*nextItems, newWorryLevel)
			} else {
				nextItems := &monkeys[monkey.throwFalse].items
				*nextItems = append(*nextItems, newWorryLevel)
			}
		}

		monkey.inspectedCount += uint64(len(monkey.items))
		monkey.items = monkey.items[:0]
	}
}

func monkeyBusiness(monkeys []*monkey, rounds int, decreaseWorryLevel bool) uint64 {
	for i := 0; i < rounds; i++ {
		round(monkeys, decreaseWorryLevel)
	}
	sort.Slice(monkeys, func(i, j int) bool {
		return monkeys[i].inspectedCount > monkeys[j].inspectedCount
	})
	return monkeys[0].inspectedCount * monkeys[1].inspectedCount
}

func main() {
	monkeysRawConcat := string(check2(os.ReadFile("input_test.txt")))
	monkeysRawLines := strings.Split(strings.TrimSpace(monkeysRawConcat), "\n")
	monkeysRaw := splitOnNewline(monkeysRawLines)
	fmt.Println(monkeyBusiness(parse(monkeysRaw), 20, true))
	fmt.Println(monkeyBusiness(parse(monkeysRaw), 800, false))
}

File: 11/11.py
from pathlib import Path
from itertools import groupby
import copy

monkeys_raw = Path('input_test.txt').read_text().splitlines()

def split(lines):
    return (list(group) for k, group in groupby(lines, lambda x: x == '') if not k)

def parse():
    for i, monkey_raw in enumerate(split(monkeys_raw)):
        op_raw = ' '.join(monkey_raw[2].split()[-3:])
        items_raw = monkey_raw[1].split(':')[1].strip().split(', ')
        monkey = {
            'items': [int(item) for item in items_raw],
            'op': eval(f'lambda old: {op_raw}'),
            'divisible_test': int(monkey_raw[3].split()[-1]),
            'throw_true': int(monkey_raw[4].split()[-1]),
            'throw_false': int(monkey_raw[5].split()[-1]),
            'inspected_count': 0,
        }
        yield monkey

monkeys = list(parse())

def round(monkeys, decrease_worry_level):
    for monkey in monkeys:
        for item in monkey['items']:
            new_worry_level = monkey['op'](item)
            if decrease_worry_level:
                new_worry_level //= 3
            if new_worry_level%monkey['divisible_test'] == 0:
                monkeys[monkey['throw_true']]['items'].append(new_worry_level)
            else:
                monkeys[monkey['throw_false']]['items'].append(new_worry_level)
        monkey['inspected_count'] += len(monkey['items'])
        monkey['items'].clear()

def monkey_business(rounds, decrease_worry_level):
    my_monkeys = copy.deepcopy(monkeys)
    for i in range(rounds):
        print(i)
        round(my_monkeys, decrease_worry_level)
    my_monkeys.sort(key=lambda x: x['inspected_count'], reverse=True)
    return my_monkeys[0]['inspected_count'] * my_monkeys[1]['inspected_count']

print(monkey_business(20, True))
print(monkey_business(10000, False))